include/rmm/device_uvector.hpp

```diff
@@ -0,0 +1,267 @@
+ * allocation, kernels, or memcpys take a CUDA stream parameter to indicate on which stream the
+ * operation will be performed.
+ *
+ * @tparam T Trivially copyable element type
+ */
+template <typename T>
+class device_uvector {
+  static_assert(std::is_trivially_copyable<T>::value,
+                "device_uvector only supports types that are trivially copyable.");
+
+ public:
+  using value_type      = T;
+  using size_type       = std::size_t;
+  using reference       = value_type&;
+  using const_reference = value_type const&;
+  using pointer         = value_type*;
+  using const_pointer   = value_type const*;
+  using iterator        = pointer;
+  using const_iterator  = const_pointer;
+
+  device_uvector()                 = default;
+  ~device_uvector()                = default;
+  device_uvector(device_uvector&&) = default;
+  device_uvector& operator=(device_uvector&&) = default;
+  // Default copying operations are deleted as they don't provide an option for specifying a stream
+  device_uvector(device_uvector const&) = delete;
+  device_uvector& operator=(device_uvector const&) = delete;
+
+  /**
+   * @brief Construct a new `device_uvector` with sufficient uninitialized storage for `size`
+   * elements.
+   *
+   * Elements are uninitialized. Reading an element before it is initialized results in undefined
+   * behavior.
+   *
+   * @param size The number of elements to allocate storage for
+   * @param stream The stream on which to perform the allocation
+   * @param mr The resource used to allocate the device storage
+   */
+  explicit device_uvector(std::size_t size,
+                          cudaStream_t stream                 = 0,
+                          rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource())
+    : _storage{elements_to_bytes(size), stream, mr}
+  {
+  }
+
+  /**
+   * @brief Construct a new device_uvector by deep copying the contents of another `device_uvector`.
+   *
+   * Elements are copied as if it were performed by `memcpy`, i.e., `T`'s copy constructor is not
+   * invoked.
+   *
+   * @param other The vector to copy from
+   * @param stream The stream on which to perform the copy
+   * @param mr The resource used to allocate device memory for the new vector
+   */
+  explicit device_uvector(device_uvector const& other,
+                          cudaStream_t stream                 = 0,
+                          rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource())
+    : _storage{other.storage, stream, mr}
+  {
+  }
+
+  /**
+   * @brief Constructs a `device_uvector` by copying the contents of a `std::vector`.
+   *
+   * @param v The host vector whose contents will be copied to device
+   * @param stream The stream on which to perform the copy
+   * @param mr The resource used to allocate device memory for the new vector
+   */
+  explicit device_uvector(std::vector<T> const& v,
```

Comment:
I think ctor actually either needs to be removed or needs to synchronize. It's far too likely to be a foot gun. Consider this simple example with undefined behavior:
```
template <typename T>
device_uvector<T> foo(cudaStream_t s){
   std::vector<T> v{1,2,3,4};
   return device_uvector<T>{v, s}; // This can return immediately before the copy is complete, which means the copy may not happen until `v` is has already been destroyed 
}
```